cclient/http/http.c

```diff
@@ -33,180 +78,438 @@ static char const* header_template =
-  return RC_OK;
+static void http_client_free(http_client_ctx_t* client_ctx) {
+  if (client_ctx) {
+    mbedtls_socket_close(client_ctx->tls);
+    free(client_ctx->tls);
+    http_buffer_free(&client_ctx->request->buffer);
+    free(client_ctx->request);
+    http_buffer_free(&client_ctx->response->buffer);
+    free(client_ctx->response);
+    http_buffer_free(&client_ctx->res_data);
+    free(client_ctx->parser);
+    free(client_ctx->parser_settings);
+    free(client_ctx);
+  }
 }
 
-/**
- * @brief sent out http request header
- *
- * @param ctx[in] mbedtls context
- * @param http_settings[in] http configurations
- * @param data_length[in] number of bytes need to send
- * @return int number of bytes actually sent
- */
-static int http_request_header(mbedtls_ctx_t* ctx, http_info_t const* http_settings, size_t data_length) {
-  char header[1024] = {};
-  sprintf(header, header_template, http_settings->path, http_settings->host, http_settings->api_version,
-          http_settings->content_type, http_settings->accept, data_length);
-  return mbedtls_socket_send(ctx, header, strlen(header));
+static http_client_ctx_t* http_client_init(void const* const service_opaque) {
+  iota_client_service_t const* const service = (iota_client_service_t const* const)service_opaque;
+  http_client_ctx_t* client = NULL;
+  bool init =
+      ((client = calloc(1, sizeof(http_client_ctx_t))) && (client->parser = calloc(1, sizeof(struct http_parser))) &&
+       (client->parser_settings = calloc(1, sizeof(struct http_parser_settings))) &&
+       (client->tls = malloc(sizeof(mbedtls_ctx_t))) && (client->request = calloc(1, sizeof(http_data_t))) &&
+       (client->response = calloc(1, sizeof(http_data_t))));
+
+  if (!init) {
+    printf("OOM\n");
+    goto error;
+  }
+  client->http_config = &service->http;
+  // TODO: configure by methods
+  client->buffer_size_rx = CCLIENT_HTTP_BUFFER_SIZE;
+  client->buffer_size_tx = CCLIENT_HTTP_BUFFER_SIZE;
+
+  init = ((client->request->buffer = http_buffer_new(client->buffer_size_tx)) &&
+          (client->response->buffer = http_buffer_new(client->buffer_size_rx)));
+
+  if (!init) {
+    printf("HTTP buffer Allocation failed\n");
+    goto error;
+  }
+
+  client->parser_settings->on_message_begin = parser_on_message_begin;
+  client->parser_settings->on_url = parser_on_url;
+  client->parser_settings->on_status = parser_on_status;
+  client->parser_settings->on_header_field = parser_on_header_field;
+  client->parser_settings->on_header_value = parser_on_header_value;
+  client->parser_settings->on_headers_complete = parser_on_headers_complete;
+  client->parser_settings->on_body = parser_on_body;
+  client->parser_settings->on_message_complete = parser_on_message_complete;
+  client->parser_settings->on_chunk_header = parser_on_chunk_header;
+  client->parser_settings->on_chunk_complete = parser_on_chunk_complete;
+  http_parser_init(client->parser, HTTP_RESPONSE);
+  client->parser->data = client;
```

Comment:
Sorry, I'm a little confused on how this works. Could you elaborate it?

Reply:
keeps the client object in the http_parser that we can use it in the http parser. 
see the parser callbacks like `parser_on_headers_complete`.